File: array/0152_maximum_product_subarray/0152_maximum_product_subarray.go
package main 

import "fmt"

func maxProduct(nums []int) int {
    minimum, maximum, res := nums[0], nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i] < 0 {
			maximum, minimum = minimum, maximum
		}
		maximum = max(nums[i], maximum*nums[i])
		minimum = min(nums[i], minimum*nums[i])
		res = max(res, maximum)
	}
	return res
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func min(a,b int) int {
    if a < b {
        return a
    }
    return b
}

func main() {
	a := []int{2,3,-2,4}
	res := maxProduct(a)
	fmt.Println(res)
}

File: array/0152_maximum_product_subarray/0152_maximum_product_subarray.py
class Solution(object):
    def maxProduct(self, A):
        global_max, local_max, local_min = float('-inf'), 1, 1
        for x in A:
            local_max, local_min = max(x, local_max * x, local_min * x), min(x, local_max * x, local_min * x)
            global_max = max(global_max, local_max)
        return global_max

A = [2,3,-2,4]

res = Solution().maxProduct(A)
print(res)